src/bin/ipa_bench/models.rs

```diff
@@ -109,57 +111,169 @@ impl SecretSharable for u64 {
+    ///
+    /// Use `u32` (`< 2^20 seconds`) to leverage simple arithmetics, but we drop the first 12 bits when serializing.
+    /// That'll save us ~1.5G with 1B rows.
+    fn offset(&self) -> Self::Offset {
+        (self.0 % Self::SECONDS_IN_EPOCH) as Self::Offset
+    }
 }
 
-#[cfg(feature = "debug")]
-impl Debug for SourceEvent {
-    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
-        write!(
-            f,
-            "SourceEvent:\n  matchkeys={:?}\n  breakdown_key={}",
-            &self.event.matchkeys, &self.breakdown_key
-        )
+impl From<u64> for EventTimestamp {
+    fn from(v: u64) -> Self {
+        EventTimestamp(v)
     }
 }
 
-#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
-pub struct TriggerEvent {
-    pub event: Event,
+impl Add for EventTimestamp {
+    type Output = Self;
 
-    /// Conversion value.
-    pub value: SecretShare,
+    fn add(self, rhs: Self) -> Self::Output {
+        EventTimestamp(self.0 + rhs.0)
+    }
 }
 
-#[cfg(feature = "debug")]
-impl Debug for TriggerEvent {
-    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
-        write!(
-            f,
-            "TriggerEvent:\n  matchkeys={:?}\n  value={:?}",
-            &self.event.matchkeys, &self.value
-        )
+impl AddAssign for EventTimestamp {
+    // #[allow(clippy::assign_op_pattern)]
+    fn add_assign(&mut self, rhs: Self) {
+        *self = *self + rhs;
+    }
+}
+
+impl PartialOrd for EventTimestamp {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(u64::from(*self).cmp(&u64::from(*other)))
+    }
+}
+
+impl From<EventTimestamp> for u64 {
+    fn from(v: EventTimestamp) -> Self {
+        v.0 % (EventTimestamp::MAX + 1)
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Event {
+    // An identifier, set in the user agent, which identifies an individual person. This must never be released (beyond
+    /// the match key provider) to any party in unencrypted form. For the purpose of this tool, however, the value is in
+    /// clear and not secret shared.
+    pub matchkey: MatchKey,
+
+    /// An identifier, specified by the report collector, to denote if a given pair of source and trigger events can be
+    /// attributed (beyond having the same match key.) If None, a trigger event will be attributed to all source events.
+    pub attribution_constraint_id: Option<Number>,
+
+    /// A timestamp in which this event is generated.
+    pub timestamp: EventTimestamp,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum GenericReport {
+    /// An event produced on websites/apps when a user interacts with an ad (i.e. impression, click).
+    Source {
+        event: Event,
+
+        /// A key, specified by the report collector, which allows for producing aggregates across many groups (or breakdowns.)
+        /// This field is only valid if is_trigger_report is [false].
```

Comment:
`is_trigger_report` is no longer valid, since this is an enum now, right?

Reply:
Thanks for catching!